### PR TITLE
[travis] basic ci tests

### PR DESCRIPTION
**Pull request by gjc13**
@jwhui need to activate travis with owner permission

### CODE REVIEW

File: third_party/lwip/port/lwipopts.h [hunk 1]
```diff
@@ -62,7 +62,7 @@
    ---------- Memory options ----------
    ------------------------------------
 */
-#define MEM_LIBC_MALLOC  1
+#define MEM_LIBC_MALLOC 1
```
Comment (gjc13) [inline third_party/lwip/port/lwipopts.h:65]: Prettify

**Comment by jwhui** [comment]
Travis is now enabled. It appears the build check is not currently passing.

File: .travis/script.sh [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+#  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+#  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+#  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+#  POSSIBILITY OF SUCH DAMAGE.
+#
+
+set -e
+set -x
+
+TOOLS_HOME=$HOME/.cache/tools
```
Comment (bukepo) [inline .travis/script.sh]: are we using this in this project?

File: .travis/script.sh [hunk 3]
```diff
@@ -0,0 +1,62 @@
+#!/bin/sh
```
Comment (bukepo) [inline .travis/script.sh]: Let's keep using bash for new scripts.

File: .travis/script.sh [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+linux-check)
+    ./script/test linux_build 
+    ;;
+
+nrf-check)
+    sudo apt-get update
+    sudo apt-get install lib32z1 || die
+    wget https://developer.arm.com/-/media/Files/downloads/gnu-rm/8-2018q4/gcc-arm-none-eabi-8-2018-q4-major-linux.tar.bz2 || die
+    tar xjf gcc-arm-none-eabi-8-2018-q4-major-linux.tar.bz2 -C /tmp || die
```
Comment (bukepo) [inline .travis/script.sh]: no need for die.

File: .travis/script.sh [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+#  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+#  POSSIBILITY OF SUCH DAMAGE.
+#
+
+set -e
+set -x
+
+TOOLS_HOME=$HOME/.cache/tools
+
+die() {
```
Comment (bukepo) [inline .travis/script.sh]: I think we don't need die with `set -e`

File: .travis/script.sh [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+    ;;
+
+linux-check)
+    ./script/test linux_build 
+    ;;
+
+nrf-check)
+    sudo apt-get update
+    sudo apt-get install lib32z1 || die
+    wget https://developer.arm.com/-/media/Files/downloads/gnu-rm/8-2018q4/gcc-arm-none-eabi-8-2018-q4-major-linux.tar.bz2 || die
```
Comment (bukepo) [inline .travis/script.sh]: create a `script/bootstrap` for this?

File: CMakeLists.txt [hunk 7]
```diff
@@ -178,8 +178,9 @@ if (${PLATFORM_NAME} STREQUAL nrf52)
     add_custom_target(ot_demo_101_hex ALL DEPENDS ot_demo_101.hex)
 endif()
 
+file(GLOB PORT_DIRS third_party/*/port)
```
Comment (bukepo) [inline CMakeLists.txt]: is it more flexible just explicitly list those directories?

File: script/test [hunk 8]
```diff
@@ -41,40 +41,19 @@ readonly COLOR_PASS='\033[0;32m'
...
-    (cd "${BUILD_DIR}" && cmake .. -DCMAKE_TOOLCHAIN_FILE=../cmake/arm-none-eabi.cmake -DPLATFORM_NAME="${PLATFORM_NAME}" && make -j${BUILD_JOBS})
-}
-
 do_clean() {
     [[ ! -d "${BUILD_DIR}" ]] || rm -rvf "${BUILD_DIR}"
+}
 
-    do_build
+do_build() {
+    do_clean
```
Comment (bukepo) [inline script/test]: Keep the original version so that we can use this for rebuild?

File: script/test [hunk 9]
```diff
@@ -41,40 +41,19 @@ readonly COLOR_PASS='\033[0;32m'
...
 
 check_pretty() {
-    pass=false
-    find src include \( -name '*.c' -o -name '*.cpp' -o -name '*.h' -o -name '*.hpp' \) | while read file; do
-        echo Checking "${file}"
-        third_party/openthread/repo/script/clang-format-check.sh "${file}" || break
-        pass=true
-    done
-    print_result "${pass}" pretty check
+    find src include third_party/*/port \( -name '*.c' -o -name '*.cpp' -o -name '*.h' -o -name '*.hpp' \) |\
```
Comment (bukepo) [inline script/test]: Will this give us which file is failing?

File: script/test [hunk 10]
```diff
@@ -100,8 +79,11 @@ main()
             clean)
                 do_clean
                 ;;
-            build)
-                do_build
+            linux_build)
```
Comment (bukepo) [inline script/test]: how about `build-linux`?

File: script/test [hunk 11]
```diff
@@ -100,8 +79,11 @@ main()
...
+                do_build -DPLATFORM_NAME=linux
+                ;;
+            nrf_build)
```
Comment (bukepo) [inline script/test]: how about `build-nrf52840`?

File: .travis.yml [hunk 12]
```diff
@@ -0,0 +1,25 @@
+language: cpp
+
+sudo: required
+dist: xenial
+
+cache:
+  directories:
+  - $HOME/.cache/tools
```
Comment (bukepo) [inline .travis.yml]: not used.

File: .travis/script.sh (hunk 2 above)
Comment (gjc13) [reply .travis/script.sh]: clang-format path

File: script/test (hunk 9 above)
Comment (gjc13) [reply script/test]: The diff output will tell us about the filename.

File: .travis.yml (hunk 12 above)
Comment (gjc13) [reply .travis.yml]: clang-format path

File: script/test [hunk 13]
```diff
@@ -41,40 +41,24 @@ readonly COLOR_PASS='\033[0;32m'
...
-    (cd "${BUILD_DIR}" && cmake .. -DCMAKE_TOOLCHAIN_FILE=../cmake/arm-none-eabi.cmake -DPLATFORM_NAME="${PLATFORM_NAME}" && make -j${BUILD_JOBS})
+    mkdir "${BUILD_DIR}"
+    (cd "${BUILD_DIR}" && cmake .. $@ && make -j${BUILD_JOBS})
 }
 
-do_clean() {
-    [[ ! -d "${BUILD_DIR}" ]] || rm -rvf "${BUILD_DIR}"
-
-    do_build
+do_clean_build() {
```
Comment (bukepo) [inline script/test]: keep the original version, we can use
`./script/test clean build` to clean and build.
`./script/test build` to rebuild if there's any error.

File: script/bootstrap_travis_arm [hunk 14]
```diff
@@ -0,0 +1,34 @@
+#!/bin/bash
```
Comment (bukepo) [inline script/bootstrap_travis_arm]: follow this file `https://github.com/openthread/openthread/blob/master/script/bootstrap`?

File: .travis/script.sh (hunk 2 above)
Comment (bukepo) [reply .travis/script.sh]: clang-format is not in this path.

File: .travis/script.sh [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+    ./script/test build_linux
+    ;;
+
+nrf-check)
+    ./script/bootstrap
+    ./script/test build_nrf52840
+    ;;
+
+*)
+    die
```
Comment (bukepo) [inline .travis/script.sh]: No die

File: script/test [hunk 16]
```diff
@@ -41,40 +41,19 @@ readonly COLOR_PASS='\033[0;32m'
...
-    (cd "${BUILD_DIR}" && cmake .. -DCMAKE_TOOLCHAIN_FILE=../cmake/arm-none-eabi.cmake -DPLATFORM_NAME="${PLATFORM_NAME}" && make -j${BUILD_JOBS})
-}
-
 do_clean() {
     [[ ! -d "${BUILD_DIR}" ]] || rm -rvf "${BUILD_DIR}"
+}
 
-    do_build
+do_build() {
+    mkdir "${BUILD_DIR}"
```
Comment (bukepo) [inline script/test]: => `[[ -d "${BUILD_DIR}" ]] || mkdir "${BUILD_DIR}"`

File: script/test [hunk 17]
```diff
@@ -41,40 +41,19 @@ readonly COLOR_PASS='\033[0;32m'
...
+    (cd "${BUILD_DIR}" && cmake .. $@ && make -j${BUILD_JOBS})
```
Comment (bukepo) [inline script/test]: `$@` => `"$@"`

**Review by bukepo — APPROVED** [review]

File: .travis.yml [hunk 18]
```diff
@@ -0,0 +1,21 @@
+language: cpp
```
Comment (jwhui) [inline .travis.yml:29]: nit: add OpenThread copyright header

**Review by jwhui — APPROVED** [review]

**Pull request merged**